crates/core/tedge_api/src/health.rs

```diff
@@ -1,12 +1,97 @@
 use std::process;
+use std::sync::Arc;
 
+use crate::mqtt_topics::ServiceTopicId;
 use mqtt_channel::Message;
 use mqtt_channel::PubChannel;
 use mqtt_channel::Topic;
 use mqtt_channel::TopicFilter;
 use serde_json::json;
 use time::OffsetDateTime;
 
+/// Encodes a valid health topic.
+///
+/// Health topics are topics on which messages about health status of services are published. To be
+/// able to send health messages, a health topic needs to be constructed for a given entity.
+// TODO: replace `Arc<str>` with `ServiceTopicId` after we're done with transition to new topics
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ServiceHealthTopic(Arc<str>);
```

Comment:
I'm not convinced by this struct which mixes unrelated things and has overlapping responsibilities with other structs.

- Is this a struct to materialize health topic structures? To build and send health messages?
- Building topics from a root prefix is the responsibility of `mqtt_schema'.
- Sending messages should be done by actors and not using a direct `PubChannel`.

Reply:
> * Is this a struct to materialize health topic structures? To build and send health messages?

Yes, the intent was to have a type wrapper for a health topic so that one has to first produce a valid health topic in order to be able to send health messages on it, in order not to send a health message on something that is not a health topic. The same of course could be done by putting guards on previously used functions to check if input is a health topic before sending the message, but since all the functions take this argument, it made sense to make it a struct and do the check only once when constructing.


> * Building topics from a root prefix is the responsibility of `mqtt_schema'.

Yeah, you're right that `new()` method on this type is probably unnecessary and it's better to use `MqttSchema` to construct it for the "new topic" case. It still needs to handle old topics though, because all the services use the same health monitoring extension and some were not yet updated to new health topics (e.g. one still have to update tests for them), so that's why for now it's just a simple string wrapper, once the transition is complete it can take `ServiceTopicId` so there would be no dependency for knowing the root prefix. The health extension actor could use it like this:

1. take as input `ServiceTopicId` of a service we need a health check for
2. Create a `ServiceHealthTopic` from `ServiceTopicId`, and use `ServiceHealthTopic` to generate health messages and make use of other health-message-related functionality
3. Use `MqttSchema` to create plain MQTT topic for service health status
4. Send MQTT publish message to MQTT actor, to publish on topic generated by `MqttSchema`, with payload generated by `ServiceHealthTopic`

with steps 3/4 perhaps subject to change because it'd be better to just have a single source for generating health messages. Will refactor this struct to work more like this.


> Sending messages should be done by actors and not using a direct `PubChannel`.

That's how it was working previously and I haven't changed it, but I can certainly improve it in this PR. For previous comment i didn't have context, but "should be done by actors instead of `PubChannel`" is very clear, so I'll make necessary modifications.

Reply:
I'm under the impression that health topics are valid only for services, and not for devices. If there were device health status topics, then I'd agree that `ServiceTopicId` doesn't make sense and to just use `EntityTopicId` since for all possible values of `EntityTopicId` we can generate a vaild health topic.
However, if health topics are only a thing for services, then when using an `EntityTopicId` we can't generate valid health topics for all values, because an `EntityTopicId` can denote either a device or a service, so a function that takes `EntityTopicId` and generates a health status topic is partial: it has to either check with an entity store if a given `EntityTopicId` is a registered service and return an error if it isn't (which is a bit of a pain because you need to pass the entity store) OR it can encode this assumption in a type that can only be constructed when the caller has ensured these conditions hold.

This intent is not clear because the "type can only be constructed when the caller has ensured these conditions hold" part is missing from this PR. Because not services are all updated, `ServiceHealthTopic` needs to handle both old and new topic scheme.

Instead of just updating all the usages with as few code changes as possible, I tried to prototype a new API, in order not to put that off for later, possibly forgetting or having to work on more features, leaving this in a suboptimal state.

We talked with Didier and came to the agreement that in feature related PRs, it's better not to add new API but do things the simplest possible way and then do PRs entirely dedicated to changing API, so that new features are not delayed by API discussions. So even If I think this is a good abstraction, I agree that perhaps I introduced it prematurely.

But since it's already done, I can either improve this abstraction, add more documentation hopefully better explaining why I think it has a purpose, or if you're still unconvinced I can remove it and just use `EntityTopicId`.

Reply:
> I'm under the impression that health topics are valid only for services, and not for devices

IIRC, I've heard comments from Reuben that health status for devices isn't quite unreasonable either. For example, when a device is being shutdown for maintenance, it might make sense for it to report that `maintenance` status as a health status. Currently, C8y doesn't support anything like that. But, that the idea itself isn't unreasonable at thin-edge level.

> it has to either check with an entity store if a given EntityTopicId is a registered service and return an error if it isn't (which is a bit of a pain because you need to pass the entity store) OR it can encode this assumption in a type that can only be constructed when the caller has ensured these conditions hold.

Yeah, it's not very unreasonable to encode an already validated info into a concrete type so that the receiving end doesn't need to revalidate it again. The key thing here is that the caller must validate this against the entity store and not just assume things purely based on the topic structure.

> This intent is not clear because the "type can only be constructed when the caller has ensured these conditions hold" part is missing from this PR.

That's probably one of the reasons why I was struggling to fully digest the usefulness of these additional abstractions like `ServiceTopicId` and `DeviceTopicId`. For something like a health status message generator, yes, this type helps in avoiding a further validation. But, in future, we may have to support health endpoints for devices as well, for the reasons mentioned above. Now, if there's some other logic where the device specific `DeviceTopicId` is useful for a receiver, I'm fine to keep it. One use-case that I can think of is the support for commands, which is currently only limited to devices. But, we haven't fully ruled out future commands support for services either.

> So even If I think this is a good abstraction, I agree that perhaps I introduced it prematurely.

Hearing your arguments, I kinda get the value it may have in future. I just don't any use-cases, **yet**. But, I don't really have any strong arguments to justify the removal of these either, especially since you've already done the heavy-lifting of introducing them.